# Quantum Circuit Training Loop

Earlier notebooks introduced:
- variational quantum circuits
- trainable rotation parameters
- quantum classifier structures

However, the parameters were still being chosen manually.

The important missing piece was that the circuit itself was not yet learning from mistakes.

This notebook introduces the optimization loop that allows quantum model parameters to gradually improve automatically through repeated updates.

# What Makes This ML

Earlier variational quantum circuits mainly demonstrated that:
- quantum circuits can contain trainable parameters
- changing rotation angles changes measurement probabilities

However, true machine learning also requires:
- predictions
- error evaluation
- parameter updates
- iterative improvement

We introduce that feedback loop.

The circuit will:
- make a prediction
- compare it with a target value
- compute error
- adjust parameters
- repeat the process

The parameter gradually moves toward values that produce better predictions.

This is similar to optimization in classical machine learning systems.

In [3]:
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
import numpy as np

# target probability
target = 0.8

# initial parameter
theta = 0.1

# learning rate
lr = 0.1

# training loop
for step in range(20):

    # create circuit
    qc = QuantumCircuit(1)

    # trainable rotation
    qc.ry(theta, 0)

    # get statevector
    state = Statevector.from_instruction(qc)

    # probability of measuring 1
    prob_1 = abs(state.data[1])**2

    # compute error
    loss = (target - prob_1)**2

    # print progress
    print(f"Step {step}")
    print(f"Theta = {theta:.4f}")
    print(f"Probability of 1 = {prob_1:.4f}")
    print(f"Loss = {loss:.6f}")
    print()

    # simple manual update rule
    if prob_1 < target:
        theta += lr
    else:
        theta -= lr

Step 0
Theta = 0.1000
Probability of 1 = 0.0025
Loss = 0.636010

Step 1
Theta = 0.2000
Probability of 1 = 0.0100
Loss = 0.624153

Step 2
Theta = 0.3000
Probability of 1 = 0.0223
Loss = 0.604768

Step 3
Theta = 0.4000
Probability of 1 = 0.0395
Loss = 0.578407

Step 4
Theta = 0.5000
Probability of 1 = 0.0612
Loss = 0.545813

Step 5
Theta = 0.6000
Probability of 1 = 0.0873
Loss = 0.507895

Step 6
Theta = 0.7000
Probability of 1 = 0.1176
Loss = 0.465699

Step 7
Theta = 0.8000
Probability of 1 = 0.1516
Loss = 0.420362

Step 8
Theta = 0.9000
Probability of 1 = 0.1892
Loss = 0.373083

Step 9
Theta = 1.0000
Probability of 1 = 0.2298
Loss = 0.325072

Step 10
Theta = 1.1000
Probability of 1 = 0.2732
Loss = 0.277516

Step 11
Theta = 1.2000
Probability of 1 = 0.3188
Loss = 0.231533

Step 12
Theta = 1.3000
Probability of 1 = 0.3663
Loss = 0.188139

Step 13
Theta = 1.4000
Probability of 1 = 0.4150
Loss = 0.148212

Step 14
Theta = 1.5000
Probability of 1 = 0.4646
Loss = 0.112472

Step 15
Theta = 1.60

# Observation

The training process began with theta = 0.1

At this stage, the qubit only rotated slightly away from |0> on the Bloch sphere.

Because of this, the probability of measuring 1 was extremely small (0.0025), very far from the target probability of 0.8 so the loss initially became quite large.

The optimization loop then repeatedly checked whether the prediction was too low or too high.

Since the probability of measuring 1 was still below the target, the circuit gradually increased theta during every iteration.

As theta increased, the qubit rotated farther away from |0>, toward |1> on the Bloch sphere increasing the probability of measuring 1.

The circuit was no longer behaving like a fixed static program but learned from feedback and changed automatically in response to prediction error.